daemon/graphdriver/zfs/zfs.go

```
@@ -0,0 +1,277 @@
+	"github.com/docker/docker/pkg/parsers"
+	zfs "github.com/mistifyio/go-zfs"
+)
+
+type ZfsOptions struct {
+	fsName    string
+	mountPath string
+}
+
+func init() {
+	graphdriver.Register("zfs", Init)
+}
+
+type Logger struct{}
+
+func (*Logger) Log(cmd []string) {
+	log.Debugf("[zfs] %s", strings.Join(cmd, " "))
+}
+
+func Init(base string, opt []string) (graphdriver.Driver, error) {
+	var err error
+	options, err := parseOptions(opt)
+	options.mountPath = base
+	if err != nil {
+		return nil, err
+	}
+
+	rootdir := path.Dir(base)
+
+	if options.fsName == "" {
+		err = checkRootdirFs(rootdir)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if _, err := exec.LookPath("zfs"); err != nil {
+		return nil, fmt.Errorf("zfs command is not available")
+	}
+
+	file, err := os.OpenFile("/dev/zfs", os.O_RDWR, 600)
+	defer file.Close()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to initialize: %v", err)
+	}
+
+	if options.fsName == "" {
+		options.fsName, err = lookupZfsDataset(rootdir)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	logger := Logger{}
+	zfs.SetLogger(&logger)
+
+	dataset, err := zfs.GetDataset(options.fsName)
+	if err != nil {
+		return nil, fmt.Errorf("Cannot open %s", options.fsName)
+	}
+
+	d := &Driver{
+		dataset: dataset,
+		options: options,
+	}
+	return graphdriver.NaiveDiffDriver(d), nil
+}
+
+func parseOptions(opt []string) (ZfsOptions, error) {
+	var options ZfsOptions
+	options.fsName = ""
+	for _, option := range opt {
+		key, val, err := parsers.ParseKeyValueOpt(option)
+		if err != nil {
+			return options, err
+		}
+		key = strings.ToLower(key)
+		switch key {
+		case "zfs.fsname":
+			options.fsName = val
+		default:
+			return options, fmt.Errorf("Unknown option %s\n", key)
+		}
+	}
+	return options, nil
+}
+
+func checkRootdirFs(rootdir string) error {
+	var buf syscall.Statfs_t
+	if err := syscall.Statfs(rootdir, &buf); err != nil {
+		fmt.Errorf("Failed to access '%s': %s", rootdir, err)
```

Should we return this error?